File: backend/app/schemas/labs.py
"""Contracts for the interactive labs.

These endpoints are *instruments*, not CRUD. Each one returns every
intermediate value the visualiser needs, because the whole point is that the
player can see the mechanism rather than the result.
"""

from __future__ import annotations

from typing import Annotated, Any

from pydantic import Field

from app.schemas.common import Schema


# ══ Transformer Lab ══════════════════════════════════════════════════════════
class AttentionRequest(Schema):
    text: Annotated[str, Field(min_length=1, max_length=400)] = "The cat sat on the mat"
    n_heads: Annotated[int, Field(ge=1, le=8)] = 4
    d_model: Annotated[int, Field(ge=16, le=256)] = 64
    #: The three toggles that are the lab's entire pedagogy.
    causal: bool = False
    scaled: bool = True
    use_positional: bool = True


class TokenOut(Schema):
    text: str
    index: int
    id: int
    is_subword: bool
    of_word: str


class AttentionHeadOut(Schema):
    head: int
    weights: list[list[float]]
    raw_scores: list[list[float]]
    entropy: list[float]
    #: Which token each row attends to most — the readable summary of a head.
    argmax: list[int]


class AttentionResponse(Schema):
    tokens: list[TokenOut]
    d_model: int
    n_heads: int
    d_head: int
    causal: bool
    scaled: bool
    heads: list[AttentionHeadOut]
    vectors: dict[str, Any] | None = None
    #: Plain-language note on what the current toggles are demonstrating.
    insight: str = ""


class TokenizeRequest(Schema):
    text: Annotated[str, Field(min_length=1, max_length=20_000)]


class TokenizeResponse(Schema):
    tokens: list[TokenOut]
    stats: dict[str, Any]
    #: What this many tokens costs across providers, at today's prices.
    cost_estimates: list[dict[str, Any]] = Field(default_factory=list)


class SamplingRequest(Schema):
    temperature: Annotated[float, Field(ge=0.01, le=3.0)] = 1.0
    top_k: Annotated[int | None, Field(ge=1, le=50)] = None
    top_p: Annotated[float | None, Field(gt=0.0, le=1.0)] = None
    logits: list[float] | None = None
    labels: list[str] | None = None


class SamplingResponse(Schema):
    labels: list[str]
    logits: list[float]
    temperature: float
    scaled_logits: list[float]
    probabilities: list[float]
    final_probabilities: list[float]
    kept: list[bool]
    tokens_kept: int
    truncation: str
    entropy: float
    argmax: str
    insight: str = ""


# ══ RAG Tower ════════════════════════════════════════════════════════════════
class RAGConfigIn(Schema):
    corpus: str = "forge-docs"
    chunk_size: Annotated[int, Field(ge=50, le=4000)] = 400
    chunk_overlap: Annotated[int, Field(ge=0, le=1000)] = 50
    respect_structure: bool = True
    top_k: Annotated[int, Field(ge=1, le=20)] = 4
    use_reranker: bool = False
    use_hybrid: bool = False
    metadata_filter: dict[str, Any] = Field(default_factory=dict)
    max_context_chars: Annotated[int, Field(ge=200, le=32_000)] = 6000
    temperature: Annotated[float, Field(ge=0.0, le=2.0)] = 0.1
    min_score: Annotated[float, Field(ge=-1.0, le=1.0)] = 0.0
    prompt_template: Annotated[str, Field(max_length=4000)] = ""


class ChunkPreviewRequest(Schema):
    text: Annotated[str, Field(min_length=1, max_length=100_000)]
    chunk_size: Annotated[int, Field(ge=50, le=4000)] = 400
    chunk_overlap: Annotated[int, Field(ge=0, le=1000)] = 50
    respect_structure: bool = True


class ChunkOut(Schema):
    index: int
    text: str
    start_char: int
    end_char: int
    chars: int
    tokens: int


class ChunkPreviewResponse(Schema):
    chunks: list[ChunkOut]
    total_chunks: int
    avg_chars: float
    avg_tokens: float
    #: Chars duplicated across chunk boundaries — the cost of overlap.
    overlap_waste_chars: int
    insight: str = ""


class RetrievedChunkOut(Schema):
    doc_slug: str
    title: str
    text: str
    score: float
    chunk_index: int
    metadata: dict[str, Any] = Field(default_factory=dict)
    rerank_score: float | None = None
    original_rank: int | None = None


class StageOut(Schema):
    name: str
    duration_ms: float
    detail: dict[str, Any] = Field(default_factory=dict)


class RAGQueryRequest(Schema):
    question: Annotated[str, Field(min_length=3, max_length=2000)]
    config: RAGConfigIn = Field(default_factory=RAGConfigIn)
    #: Supplying this scores the run against a golden case immediately.
    expected_answer: str = ""
    relevant_doc_ids: list[str] = Field(default_factory=list)
    save_experiment: bool = True


class RAGQueryResponse(Schema):
    question: str
    answer: str
    retrieved: list[RetrievedChunkOut]
    config: dict[str, Any]
    stages: list[StageOut]
    prompt: str
    total_ms: float
    token_usage: dict[str, int]
    cost_usd: float
    simulated: bool
    metrics: dict[str, float] = Field(default_factory=dict)
    experiment_id: str | None = None
    #: What the numbers are telling you, in words.
    diagnosis: list[str] = Field(default_factory=list)


class RAGCorpusOut(Schema):
    corpus: str
    documents: int
    chunks: int
    embedding_model: str | None
    total_tokens: int
    titles: list[str] = Field(default_factory=list)


class RAGExperimentOut(Schema):
    id: str
    question: str
    chunk_size: int
    chunk_overlap: int
    top_k: int
    use_reranker: bool
    use_hybrid: bool
    embedding_model: str
    answer: str
    metrics: dict[str, Any]
    latency_ms: float
    created_at: str


class RAGCompareResponse(Schema):
    experiments: list[RAGExperimentOut]
    #: Metric deltas between the two most recent runs.
    delta: dict[str, float] = Field(default_factory=dict)
    verdict: str = "hold"
    reasons: list[str] = Field(default_factory=list)


# ══ Agent Factory ════════════════════════════════════════════════════════════
class GraphNodeOut(Schema):
    id: str
    label: str
    kind: str
    description: str = ""
    interrupt_before: bool = False
    is_entry: bool = False


class GraphEdgeOut(Schema):
    source: str
    target: str
    kind: str
    label: str = ""
    description: str = ""


class GraphDiagramOut(Schema):
    name: str
    entry: str | None
    nodes: list[GraphNodeOut]
    edges: list[GraphEdgeOut]
    #: Every cycle. Not a bug — but every one needs a termination argument.
    cycles: list[list[str]] = Field(default_factory=list)
    info: dict[str, Any] = Field(default_factory=dict)


class AgentRunRequest(Schema):
    graph: str = "support_agent"
    question: Annotated[str, Field(min_length=1, max_length=2000)]
    recursion_limit: Annotated[int, Field(ge=1, le=60)] = 25
    initial_state: dict[str, Any] = Field(default_factory=dict)
    save_run: bool = True


class StepOut(Schema):
    step: int
    node: str
    status: str
    duration_ms: float
    state_diff: dict[str, Any]
    state_after: dict[str, Any]
    edge_taken: str | None = None
    edge_reason: str | None = None
    error: str | None = None
    tokens: int = 0
    cost_usd: float = 0.0


class AgentRunResponse(Schema):
    run_id: str | None = None
    graph: str
    status: str
    final_state: dict[str, Any]
    history: list[StepOut]
    halted_reason: str | None = None
    interrupted_at: str | None = None
    steps: int
    total_ms: float
    node_visits: dict[str, int]
    looped: bool
    diagnosis: list[str] = Field(default_factory=list)


class AgentResumeRequest(Schema):
    run_id: str
    approved: bool = True
    note: Annotated[str, Field(max_length=2000)] = ""


# ══ Mentor & content agent ═══════════════════════════════════════════════════
class MentorAskRequest(Schema):
    question: Annotated[str, Field(min_length=3, max_length=4000)]
    concept_slug: str | None = None
    challenge_slug: str | None = None
    code: Annotated[str | None, Field(max_length=20_000)] = None
    error: Annotated[str | None, Field(max_length=4000)] = None
    tier: Annotated[int, Field(ge=1, le=10)] = 5
    attempts: Annotated[int, Field(ge=0, le=100)] = 0
    hints_used: Annotated[int, Field(ge=0, le=20)] = 0
    persona: str = "senior_engineer"


class MentorResponseOut(Schema):
    level: str
    message: str
    question_back: str | None = None
    next_level: str | None = None
    reveals_answer: bool
    simulated: bool
    tokens: int = 0
    cost_usd: float = 0.0
    #: The mentor's own graph execution — the tool teaching its own mechanism.
    trace: list[dict[str, Any]] = Field(default_factory=list)


class SocraticRequest(Schema):
    concept_slug: str
    player_answer: Annotated[str, Field(min_length=1, max_length=20_000)]
    tier: Annotated[int, Field(ge=1, le=10)] = 5


class SocraticResponse(Schema):
    question: str
    targets_misconception: str
    good_answer_contains: list[str]
    tier: int
    simulated: bool


class GenerateQuestionRequest(Schema):
    concept_slug: str
    tier: Annotated[int, Field(ge=1, le=10)] = 5
    kind: str = "explain"
    count: Annotated[int, Field(ge=1, le=5)] = 1


class GeneratedQuestionOut(Schema):
    prompt: str
    kind: str
    tier: int
    context: str = ""
    options: list[dict[str, Any]] = Field(default_factory=list)
    expected_answer: str = ""
    ideal_senior_answer: str = ""
    common_wrong_answer: str = ""
    rubric: list[dict[str, Any]] = Field(default_factory=list)
    hints: list[str] = Field(default_factory=list)
    followups: list[str] = Field(default_factory=list)
    #: Non-empty means this was rejected and must not be served as content.
    issues: list[str] = Field(default_factory=list)
    is_usable: bool = False
    simulated: bool = False


class GenerateQuestionResponse(Schema):
    concept_slug: str
    concept_title: str
    generated: list[GeneratedQuestionOut]
    accepted: int
    rejected: int
    #: Why generated content is held to the same bar as hand-authored content.
    validation_note: str = ""


# ══ Evaluation Lab ═══════════════════════════════════════════════════════════
class EvalCaseIn(Schema):
    question: str
    relevant_doc_ids: list[str] = Field(default_factory=list)
    expected_answer: str = ""


class EvalRunRequest(Schema):
    name: Annotated[str, Field(min_length=1, max_length=120)] = "experiment"
    config: RAGConfigIn = Field(default_factory=RAGConfigIn)
    cases: list[EvalCaseIn] = Field(default_factory=list)
    #: Use the built-in golden set when no cases are supplied.
    use_golden_set: bool = True
    use_llm_judge: bool = False
    baseline_evaluation_id: str | None = None


class EvalCaseResult(Schema):
    question: str
    answer: str
    passed: bool
    metrics: dict[str, float]
    retrieved_titles: list[str] = Field(default_factory=list)
    unsupported_claims: list[str] = Field(default_factory=list)


class EvalRunResponse(Schema):
    evaluation_id: str | None = None
    name: str
    cases_total: int
    cases_passed: int
    metrics: dict[str, float]
    per_case: list[EvalCaseResult]
    verdict: str
    reasons: list[str] = Field(default_factory=list)
    baseline: dict[str, float] | None = None
    judged_by: str = "deterministic"

File: frontend/src/types/labs.ts
/**
 * Types for the interactive labs.
 *
 * Kept in their own module rather than appended to `api.ts` because the labs
 * are instruments, not CRUD: their responses carry every intermediate value the
 * visualiser needs (raw scores, per-stage timings, state diffs), so the shapes
 * are both larger and shaped by rendering needs rather than by domain entities.
 *
 * Same drift guard applies — `src/test/contract.test.ts` validates every path
 * these are used on against the committed OpenAPI schema.
 */

// ── Transformer Lab ─────────────────────────────────────────────────────────
export interface LabToken {
  text: string;
  index: number;
  id: number;
  is_subword: boolean;
  of_word: string;
}

export interface AttentionHead {
  head: number;
  /** [query][key] — rows sum to 1. This is the matrix the heatmap draws. */
  weights: number[][];
  /** Pre-softmax Q·Kᵀ(/√d). Shown so scaling stops being an abstract claim. */
  raw_scores: number[][];
  /** Per-query-row entropy: low = sharp/confident, high = diffuse. */
  entropy: number[];
  argmax: number[];
}

export interface AttentionResponse {
  tokens: LabToken[];
  d_model: number;
  n_heads: number;
  d_head: number;
  causal: boolean;
  scaled: boolean;
  heads: AttentionHead[];
  vectors: {
    embedding: number[][];
    positional: number[][] | null;
    combined: number[][];
    shown_dimensions: number;
  } | null;
  insight: string;
}

export interface AttentionRequest {
  text?: string;
  n_heads?: number;
  d_model?: number;
  causal?: boolean;
  scaled?: boolean;
  use_positional?: boolean;
}

export interface TokenizeResponse {
  tokens: LabToken[];
  stats: {
    tokens: number;
    words: number;
    characters: number;
    tokens_per_word: number;
    chars_per_token: number;
    subword_splits: number;
  };
  cost_estimates: {
    model: string;
    /** This one prompt, once. Reads as free — which is the trap. */
    cost_per_request: number;
    /** The same prompt at a million calls. This is the number that bites. */
    input_cost_per_1m_requests: number;
  }[];
}

export interface SamplingRequest {
  temperature?: number;
  top_k?: number | null;
  top_p?: number | null;
  logits?: number[];
  labels?: string[];
}

export interface SamplingResponse {
  labels: string[];
  logits: number[];
  temperature: number;
  scaled_logits: number[];
  /** After temperature, before truncation. */
  probabilities: number[];
  /** After truncation and renormalisation — what actually gets sampled. */
  final_probabilities: number[];
  kept: boolean[];
  tokens_kept: number;
  truncation: string;
  entropy: number;
  argmax: string;
  insight: string;
}

// ── RAG Tower ───────────────────────────────────────────────────────────────
export interface RAGConfig {
  corpus?: string;
  chunk_size?: number;
  chunk_overlap?: number;
  respect_structure?: boolean;
  top_k?: number;
  use_reranker?: boolean;
  use_hybrid?: boolean;
  metadata_filter?: Record<string, unknown>;
  max_context_chars?: number;
  temperature?: number;
  min_score?: number;
  prompt_template?: string;
}

export interface ChunkPreviewResponse {
  chunks: {
    index: number;
    text: string;
    start_char: number;
    end_char: number;
    chars: number;
    tokens: number;
  }[];
  total_chunks: number;
  avg_chars: number;
  avg_tokens: number;
  /** Characters duplicated across boundaries — what overlap actually costs. */
  overlap_waste_chars: number;
  insight: string;
}

export interface RetrievedChunk {
  doc_slug: string;
  title: string;
  text: string;
  score: number;
  chunk_index: number;
  metadata: Record<string, unknown>;
  rerank_score: number | null;
  /** Position before reranking — the delta is the reranker's whole value. */
  original_rank: number | null;
}

export interface PipelineStage {
  name: string;
  duration_ms: number;
  detail: Record<string, unknown>;
}

export interface RAGQueryResponse {
  question: string;
  answer: string;
  retrieved: RetrievedChunk[];
  config: Record<string, unknown>;
  stages: PipelineStage[];
  prompt: string;
  total_ms: number;
  token_usage: Record<string, number>;
  cost_usd: number;
  simulated: boolean;
  metrics: Record<string, number>;
  experiment_id: string | null;
  diagnosis: string[];
}

export interface RAGCorpus {
  corpus: string;
  documents: number;
  chunks: number;
  embedding_model: string | null;
  total_tokens: number;
  titles: string[];
}

export interface RAGExperiment {
  id: string;
  question: string;
  chunk_size: number;
  chunk_overlap: number;
  top_k: number;
  use_reranker: boolean;
  use_hybrid: boolean;
  embedding_model: string;
  answer: string;
  metrics: Record<string, number>;
  latency_ms: number;
  created_at: string;
}

export interface RAGCompareResponse {
  experiments: RAGExperiment[];
  delta: Record<string, number>;
  verdict: string;
  reasons: string[];
}

export interface GoldenCase {
  question: string;
  relevant: string[];
  expected: string;
}

// ── Agent Factory ───────────────────────────────────────────────────────────
export interface GraphSummary {
  slug: string;
  title: string;
  teaches: string;
  difficulty: number;
  /** Deliberately broken graphs exist to be debugged, not avoided. */
  broken: boolean;
  watch_for: string;
  nodes: number;
  edges: number;
  cycles: number;
}

export interface GraphNode {
  id: string;
  label: string;
  kind: string;
  description: string;
  interrupt_before: boolean;
  is_entry: boolean;
}

export interface GraphEdge {
  source: string;
  target: string;
  kind: string;
  label: string;
  description: string;
}

export interface GraphDiagram {
  name: string;
  entry: string | null;
  nodes: GraphNode[];
  edges: GraphEdge[];
  /** Not bugs — but each one needs a termination argument. */
  cycles: string[][];
  info: Record<string, unknown>;
}

export interface AgentStep {
  step: number;
  node: string;
  status: string;
  duration_ms: number;
  /** Only what this node changed — the readable unit of agent debugging. */
  state_diff: Record<string, unknown>;
  state_after: Record<string, unknown>;
  edge_taken: string | null;
  edge_reason: string | null;
  error: string | null;
  tokens: number;
  cost_usd: number;
}

export interface AgentRunResponse {
  run_id: string | null;
  graph: string;
  status: string;
  final_state: Record<string, unknown>;
  history: AgentStep[];
  halted_reason: string | null;
  interrupted_at: string | null;
  steps: number;
  total_ms: number;
  node_visits: Record<string, number>;
  looped: boolean;
  diagnosis: string[];
}

// ── Mentor & content agent ──────────────────────────────────────────────────
export type MentorLevel = 'nudge' | 'hint' | 'concept' | 'partial' | 'full';

export interface MentorAskRequest {
  question: string;
  concept_slug?: string | null;
  challenge_slug?: string | null;
  code?: string | null;
  error?: string | null;
  tier?: number;
  attempts?: number;
  hints_used?: number;
  persona?: string;
}

export interface MentorResponse {
  level: MentorLevel;
  message: string;
  question_back: string | null;
  next_level: string | null;
  reveals_answer: boolean;
  simulated: boolean;
  tokens: number;
  cost_usd: number;
  /** The mentor's own graph run — the tool demonstrating its own mechanism. */
  trace: { node: string; [key: string]: unknown }[];
}

export interface SocraticResponse {
  question: string;
  targets_misconception: string;
  good_answer_contains: string[];
  tier: number;
  simulated: boolean;
}

export interface GeneratedQuestion {
  prompt: string;
  kind: string;
  tier: number;
  context: string;
  options: { id?: string; text: string; correct?: boolean; why?: string }[];
  expected_answer: string;
  ideal_senior_answer: string;
  common_wrong_answer: string;
  rubric: { point: string; keywords: string[]; weight: number }[];
  hints: string[];
  followups: string[];
  /** Non-empty means rejected — it must not be served as content. */
  issues: string[];
  is_usable: boolean;
  simulated: boolean;
}

export interface GenerateQuestionResponse {
  concept_slug: string;
  concept_title: string;
  generated: GeneratedQuestion[];
  accepted: number;
  rejected: number;
  validation_note: string;
}

// ── Evaluation Lab ──────────────────────────────────────────────────────────
export interface EvalCaseResult {
  question: string;
  answer: string;
  passed: boolean;
  metrics: Record<string, number>;
  retrieved_titles: string[];
  unsupported_claims: string[];
}

export interface EvalRunResponse {
  evaluation_id: string | null;
  name: string;
  cases_total: number;
  cases_passed: number;
  metrics: Record<string, number>;
  per_case: EvalCaseResult[];
  verdict: string;
  reasons: string[];
  baseline: Record<string, number> | null;
  judged_by: string;
}

export interface LabStatus {
  llm: { provider: string; simulated: boolean; configured: boolean };
  embeddings: { model: string; simulated: boolean };
  note: string;
}
